components/chat_widget/src/components/ocs-chat/ocs-chat.tsx

```diff
@@ -321,13 +348,99 @@ export class OcsChat {
+      if (!sf.error && !sf.uploaded) {
+        return { ...sf, error: errorMessage };
+      }
+      return sf;
+    });
+  }
+
+  private async uploadFiles(): Promise<number[]> {
+    if (this.selectedFiles.length === 0 || !this.sessionId || !this.allowAttachments) {
+      return [];
+    }
+
+    this.isUploadingFiles = true;
+    const uploadedIds: number[] = [];
+
+    try {
+      const formData = new FormData();
+
+      // Add all files to form data
+      for (const selectedFile of this.selectedFiles) {
+        if (!selectedFile.error && !selectedFile.uploaded) {
+          formData.append('files', selectedFile.file);
+        } else if (selectedFile.uploaded) {
+          uploadedIds.push(selectedFile.uploaded.id);
+        }
+      }
+
+      // Add user ID and name to the form data
+      const userId = this.getOrGenerateUserId();
+      formData.append('participant_remote_id', userId);
+      if (this.userName) {
+        formData.append('participant_name', this.userName);
+      }
+
+      // Only upload if there are new files
+      if (formData.has('files')) {
+        const response = await fetch(`${this.getApiBaseUrl()}/api/chat/${this.sessionId}/upload/`, {
+          method: 'POST',
+          body: formData,
+        });
+
+        if (!response.ok) {
+          const errorData = await response.json();
+          const errorMessage = errorData.error || 'Failed to upload files';
+          this.markPendingFilesWithError(errorMessage);
+          return uploadedIds;
+        }
+
+        const data = await response.json();
+
+        // Update selected files with upload results
+        let fileIndex = 0;
+        this.selectedFiles = this.selectedFiles.map(sf => {
+          if (!sf.error && !sf.uploaded) {
+            return { ...sf, uploaded: data.files[fileIndex++] };
+          }
+          return sf;
+        });
+        uploadedIds.push(...data.files.map((f: UploadedFile) => f.id));
+      }
+
+      return uploadedIds;
+    } catch (error) {
+      const errorMessage = error instanceof Error ? error.message : 'Failed to upload files';
+      this.markPendingFilesWithError(errorMessage);
+      return uploadedIds;
+    } finally {
+      this.isUploadingFiles = false;
+    }
+  }
+
   private async sendMessage(message: string): Promise<void> {
     if (!this.sessionId || !message.trim()) return;
 
     // Hide starter questions on any user interaction
     this.showStarterQuestions = false;
 
     try {
+      let attachmentIds: number[] = [];
+      if (this.allowAttachments && this.selectedFiles.length > 0) {
+        attachmentIds = await this.uploadFiles();
+
+        // Check if any files have errors after upload attempt
+        const hasErrors = this.selectedFiles.some(sf => sf.error);
+        if (hasErrors) {
+          // Don't send the message, let user fix file issues first
+          this.error = 'Please remove or fix file errors before sending your message.';
```

Comment:
Setting the error here will hide the chat window entirely. If we're going to use this field we need to change how it is rendered.

But perhaps we can do that in a followup PR.

Reply:
https://github.com/dimagi/open-chat-studio/issues/2063